### PR TITLE
[Containerapp] `az containerapp create/up`: Make `--target-port` optional

### PR DESCRIPTION
**Pull request by Tratcher**
**Related command**
`az containerapp up --name port0verify2 --resource-group port0verify-rg --ingress external --image mcr.microsoft.com/azuredocs/containerapps-helloworld:latest`
`az containerapp create --name port0verify --resource-group WestCentralUsVerify --ingress external --image mcr.microsoft.com/azuredocs/containerapps-helloworld:latest --environment port0verify-env`

**Description**
The --target-port field is now optional so the validation needs to be updated accordingly.

**Testing Guide**
Before this change, creating an app without the --target-port field causes a warning `No ingress provided, defaulting to port 80. Try `az containerapp up --ingress external --target-port <port>` to set a custom port.`. Now that warning has been removed. Other commands may have reported the error `Usage error: must specify --target-port with --ingress` which has also been removed.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Container App

**Comment by bgashirabake** [comment]
hi @Tratcher ...i remember HISTORY.rst and test files being an imperative for cli-extensions (and maybe cli-core?) PRs...could you double-check here: https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules

**Comment by bgashirabake** [comment]
> hi @Tratcher ...i remember HISTORY.rst and test files being an imperative for cli-extensions (and maybe cli-core?) PRs...could you double-check here: https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules

you have changes to the decorator and as it's a central code piece, you may have to run all relevant tests and include recording files

File: src/azure-cli/azure/cli/command_modules/containerapp/_validators.py [hunk 1]
```diff
@@ -132,18 +132,11 @@ def validate_target_port(namespace):
...
-                raise ValidationError("Usage error: must specify --target-port with --ingress")
-
-
 def validate_allow_insecure(namespace):
     if "create" in namespace.command.lower():
         if namespace.allow_insecure:
-            if not namespace.ingress or not namespace.target_port:
-                raise ValidationError("Usage error: must specify --ingress and --target-port with --allow-insecure")
+            if not namespace.ingress:
+                raise ValidationError("Usage error: must specify --ingress with --allow-insecure")
```
Comment (Greedygre) [inline src/azure-cli/azure/cli/command_modules/containerapp/_validators.py:139]: `validate_allow_insecure` used by `az containerapp create`, Do you also want to update `az containerapp create` to make target-port optional?

**Comment by Greedygre** [comment]
Please add all related commands to your change in the PR's description.
Add tests to cover your change.

**Comment by Greedygre** [comment]
> The --target-port field is now optional so the validation needs to be updated accordingly.

@Tratcher 
When --target-port optional to all regions?

**Comment by Tratcher** [comment]
> > The --target-port field is now optional so the validation needs to be updated accordingly.
> 
> @Tratcher When --target-port optional to all regions?

The service changes have already been deployed to all regions.

File: src/azure-cli/azure/cli/command_modules/containerapp/_validators.py (hunk 1 above)
Comment (Tratcher) [reply src/azure-cli/azure/cli/command_modules/containerapp/_validators.py:139]: Yes, it's also optional for `az containerapp create`

**Comment by Tratcher** [comment]
Tests updated and passing. Feedback addressed. Please review.

`Check the Format of Pull Request Title and Content` should pass on next run.

**Comment by Greedygre** [comment]
Recommend title:
[Container app] `az containerapp create/update/up`: Make `--target-port` optional

**Comment by Greedygre** [comment]
@Tratcher 
1.please get approval from your team who could be responsible for your business logic.
2.please rerun all tests relate to the command and logic you change.
Thanks

**Comment by Tratcher** [comment]
> Recommend title: [Container app] `az containerapp create/update/up`: Make `--target-port` optional

@Greedygre update does not have this option. I've updated the title accordingly.

> 2.please rerun all tests relate to the command and logic you change.

I have, locally and on the CI. Do you mean that I should check in new recording files for every containerapps test?

File: src/azure-cli/azure/cli/command_modules/containerapp/_params.py [hunk 2]
```diff
@@ -273,7 +273,7 @@ def load_arguments(self, _):
 
     with self.argument_context('containerapp ingress') as c:
         c.argument('allow_insecure', arg_type=get_three_state_flag(), help='Allow insecure connections for ingress traffic.')
-        c.argument('type', validator=validate_ingress, arg_type=get_enum_type(['internal', 'external']), help="The ingress type.")
+        c.argument('type', arg_type=get_enum_type(['internal', 'external']), help="The ingress type.")
         c.argument('transport', arg_type=get_enum_type(['auto', 'http', 'http2', 'tcp']), help="The transport protocol used for ingress traffic.")
         c.argument('target_port', type=int, validator=validate_target_port, help="The application port used for ingress traffic.")
```
Comment (smzeng) [inline src/azure-cli/azure/cli/command_modules/containerapp/_params.py:278]: do for validate target port here do we need to also allow 0?

**Review by smzeng — APPROVED** [review]

**Comment by yonzhan** [comment]
Please fix CI issues

**Comment by Tratcher** [comment]
> Please fix CI issues

@yonzhan I already updated the title so `(Check the Format of Pull Request Title and Content)` should pass on re-run. Can you re-trigger it? I can't.

**Review by Greedygre — APPROVED** [review]
Should change `az containerapp ingress enable --target-port` too? You can change with a separate PR if need.

File: src/azure-cli/azure/cli/command_modules/containerapp/_constants.py [hunk 3]
```diff
@@ -74,7 +74,7 @@
   - push: ["$Registry/{{image_name}}"]
     timeout: 1800
 """
-DEFAULT_PORT = 8080  # used for no dockerfile scenario; not the hello world image
+DEFAULT_PORT = 0  # used for no dockerfile scenario; not the hello world image. Auto detect the port
```
Comment (yanzhudd) [inline src/azure-cli/azure/cli/command_modules/containerapp/_constants.py:77]: May I ask if this change would bring a breaking change to users?

Comment (Tratcher) [reply src/azure-cli/azure/cli/command_modules/containerapp/_constants.py:77]: It shouldn't be, we'll detect the port for them.

**Comment by Tratcher** [comment]
> Should change `az containerapp ingress enable --target-port` too? You can change with a separate PR if need.

@Greedygre is _sdk_models.py the right place to change that? Updated. I didn't see it change the help output locally.

**Comment by Greedygre** [comment]
> > Should change `az containerapp ingress enable --target-port` too? You can change with a separate PR if need.
> 
> @Greedygre is _sdk_models.py the right place to change that? Updated. I didn't see it change the help output locally.

@Tratcher 
The --target-port argument is required or not is base on argument in function `enable_ingress`, the `target_port` is required inthe function.
Change --target-port to no required is change it from:
`def enable_ingress(cmd, name, resource_group_name, type, target_port, transport="auto", exposed_port=None, allow_insecure=False, disable_warnings=False, no_wait=False):  # pylint: disable=redefined-builtin`

to:
`def enable_ingress(cmd, name, resource_group_name, type, target_port=None, transport="auto", exposed_port=None, allow_insecure=False, disable_warnings=False, no_wait=False):  # pylint: disable=redefined-builtin`

**Comment by Greedygre** [comment]
The `_sdk_models.py` is only used for `--yaml` for Serialization and Deserialization. Please revert the change in it, not relate to this.

File: src/azure-cli/azure/cli/command_modules/containerapp/custom.py [hunk 4]
```diff
@@ -2444,14 +2444,16 @@ def enable_ingress(cmd, name, resource_group_name, type, target_port, transport=
             external_ingress = True
 
     ingress_def = None
-    if target_port is not None and type is not None:
+    if type is not None:
```
Comment (Greedygre) [inline src/azure-cli/azure/cli/command_modules/containerapp/custom.py:2447]: Please add test to cover this change.

Comment (Greedygre) [reply src/azure-cli/azure/cli/command_modules/containerapp/custom.py:2447]: Otherwise LGTM.

Comment (Tratcher) [reply src/azure-cli/azure/cli/command_modules/containerapp/custom.py:2447]: Added

**Review by Greedygre — APPROVED** [review]

**Review by yanzhudd — APPROVED** [review]

**Pull request merged**